models/classes/counter/CounterService.php

```diff
@@ -0,0 +1,195 @@
+    }
+
+    /**
+     * @param string $eventFqcn
+     *
+     * @throws InvalidServiceManagerException|CounterServiceException
+     * @throws \oat\oatbox\service\ServiceNotFoundException
+     * @throws \common_Exception
+     */
+    public function detach(string $eventFqcn): void
+    {
+        $eventOptions = $this->getOption(self::OPTION_EVENTS);
+
+        if (!array_key_exists($eventFqcn, $eventOptions)) {
+            $msg = "No event with FQCN '${eventFqcn}' already attached for counting.";
+            throw new CounterServiceException($msg, CounterServiceException::CODE_UNKNOWN_EVENT_TYPE);
+        }
+
+        /** @var EventManager $eventManager */
+        $eventManager = $this->getServiceManager()->get(EventManager::SERVICE_ID);
+        $eventManager->detach($eventFqcn, [self::SERVICE_ID, 'increment']);
+        $this->getServiceManager()->register(EventManager::SERVICE_ID, $eventManager);
+
+        unset($eventOptions[$eventFqcn]);
+
+        $this->setOption(self::OPTION_EVENTS, $eventOptions);
+
+        $this->logDebug("CounterService not listening anymore to events with FQCN '${eventFqcn}'.");
+    }
+
+    /**
+     * @param Event $event
+     * @throws CounterServiceException
+     */
+    public function increment(Event $event): void
+    {
+        $eventOptions = $this->getOption(self::OPTION_EVENTS);
+        $eventFqcn = get_class($event);
+
+        if (!array_key_exists($eventFqcn, $eventOptions)) {
+            throw new CounterServiceException(
+                "Configuration Violation. No event with FQCN '${eventFqcn}' registered while incrementing."
+            );
+        }
+
+        $eventOption = $eventOptions[$eventFqcn];
+        $keyCallbackMethod = $eventOption[self::OPTION_CALLBACK];
+        $keyCallbackValue = null;
+
+        if (!empty($keyCallbackMethod)) {
+            $keyCallbackValue = $event->$keyCallbackMethod();
+        }
+
+        $this->getPersistence()->incr($this->buildKey($eventFqcn, $keyCallbackValue));
+    }
+
+    /**
+     * @param string $eventFqcn
+     * @param int|null $value
+     * @param string|null $suffix
+     * @throws CounterServiceException
+     * @throws common_Exception
+     */
+    public function reset(string $eventFqcn, ?int $value = 0, ?string $suffix = null): void
+    {
+        $this->getPersistence()->set($this->buildKey($eventFqcn, $suffix), $value);
+    }
+
+    /**
+     * @param string $eventFqcn
+     * @param string|null $keyCallbackValue
+     * @return int
+     * @throws CounterServiceException
+     */
+    public function get(string $eventFqcn, ?string $keyCallbackValue = null): int
+    {
+        return (int)$this->getPersistence()->get($this->buildKey($eventFqcn, $keyCallbackValue));
+    }
+
+    /**
+     * @param string $eventFqcn
+     * @param string|null $keySuffix
+     * @return string
+     */
+    protected function buildKey(string $eventFqcn, ?string $keySuffix = null): string
```

Comment:
Keep least privilege in mind. It is very important to avoid extra maintenance costs.

https://github.com/oat-sa/tao-core/pull/3101/files#r704079172

Unnecessary annotation, we already have typehint and return types

```suggestion
    private function buildKey(string $eventFqcn, ?string $keySuffix = null): string
```